src/common.ts

```diff
@@ -72,9 +72,10 @@ export const enum BindingKind {
   Const = 1 << 4,
   Class = 1 << 5,
   FunctionLexical = 1 << 6,
-  FunctionStatement = 1 << 7,
-  CatchPattern = 1 << 8,
-  CatchIdentifier = 1 << 9,
+  AsyncFunctionLexical = (1 << 7) | FunctionLexical,
```

Comment:
Not sure if this is the best way to add AsyncFunctionLexical.

Reply:
I'm not familiar with the parser, but do we have the information about declared functions in scope? 

I saw there are failing tests for generator functions, will we add more like this?

Reply:
Thx!
Generator function is new too. Should not carry legacy behaviour either. I will fix that.

Reply:
I mean shouldn't we already have the declared functions, can't we check if function names are duplicates directly?

Reply:
All normal/async/generator functions are functions. We need to check the duplicated bindings in scope are normal function, not async nor generator.

Reply:
Aren't they already parsed, and we know it's an async function or not? 

Something like

```js
if (
  duplicatedFunctions.some(fn =>
    !fn.async
    && !fn.generator
    && fn.id.value === name)
  )
) {
  // ..
}
```

won't work? Or only because we avoid such test for performance?

Reply:
Our lexical scope doesn't save the node (identifier, function, ...) itself, it only saves a BindingKind. There is not even start/end/position saved.

The scope is a mapping like `{ "#funcName" : BindingKind.FunctionLexical }`.

The BindingKind is the only thing we can read when checking duplication.

Reply:
Thanks for the explanation.